From f3b52926da8f7659e6805779db5f25fa624dfbc9 Mon Sep 17 00:00:00 2001
From: Jammy Huang <jammy_huang@aspeedtech.com>
Date: Wed, 12 Jan 2022 16:29:13 +0800
Subject: [PATCH 24/27] media: aspeed: Fix kanic, nobody cared irq

Video engine sometimes asserts interrupts that the driver hasn't
enabled. Instead of playing whack-a-mole with each one individually,
we can instead just blanketignore everything we haven't explicitly
enabled.

Signed-off-by: Jammy Huang <jammy_huang@aspeedtech.com>
Change-Id: I52aac8ea4c2fe4dbacbf1a8d1ef3f038d0a7eb21
---
 drivers/media/platform/aspeed-video.c | 12 ++----------
 1 file changed, 2 insertions(+), 10 deletions(-)

diff --git a/drivers/media/platform/aspeed-video.c b/drivers/media/platform/aspeed-video.c
index a98a307d7230..b61de26ba990 100644
--- a/drivers/media/platform/aspeed-video.c
+++ b/drivers/media/platform/aspeed-video.c
@@ -744,6 +744,8 @@ static irqreturn_t aspeed_video_irq(int irq, void *arg)
 	struct aspeed_video *video = arg;
 	u32 sts = aspeed_video_read(video, VE_INTERRUPT_STATUS);
 
+	sts &= aspeed_video_read(video, VE_INTERRUPT_CTRL);
+
 	v4l2_dbg(2, debug, &video->v4l2_dev, "irq sts=%#x %s%s%s%s\n", sts,
 		 sts & VE_INTERRUPT_MODE_DETECT_WD ? ", unlock" : "",
 		 sts & VE_INTERRUPT_MODE_DETECT ? ", lock" : "",
@@ -832,16 +834,6 @@ static irqreturn_t aspeed_video_irq(int irq, void *arg)
 			aspeed_video_start_frame(video);
 	}
 
-	/*
-	 * CAPTURE_COMPLETE and FRAME_COMPLETE interrupts come even when these
-	 * are disabled in the VE_INTERRUPT_CTRL register so clear them to
-	 * prevent unnecessary interrupt calls.
-	 */
-	if (sts & VE_INTERRUPT_CAPTURE_COMPLETE)
-		sts &= ~VE_INTERRUPT_CAPTURE_COMPLETE;
-	if (sts & VE_INTERRUPT_FRAME_COMPLETE)
-		sts &= ~VE_INTERRUPT_FRAME_COMPLETE;
-
 	return sts ? IRQ_NONE : IRQ_HANDLED;
 }
 
-- 
2.17.1

